PurchasesCoreSwift/Identity/PurchaserInfoManager.swift

```diff
@@ -0,0 +1,191 @@
+                    completion(infoFromCache, nil)
+                }
+            }
+        }
+
+        // Prevent calling completion twice.
+        let fetchPurchaserInfoCompletion = completionCalled ? nil : completionBlock
+
+        systemInfo.isApplicationBackgrounded { isAppBackgrounded in
+            self.fetchAndCachePurchaserInfoIfStale(appUserID: appUserID,
+                                                   isAppBackgrounded: isAppBackgrounded,
+                                                   completion: fetchPurchaserInfoCompletion)
+        }
+    }
+
+    @objc(cachedPurchaserInfoForAppUserID:)
+    public func cachedPurchaserInfo(appUserID: String) -> PurchaserInfo? {
+        guard let purchaserInfoData = deviceCache.cachedPurchaserInfoData(appUserID: appUserID) else {
+            return nil
+        }
+
+        do {
+            let maybeInfoDict = try JSONSerialization.jsonObject(with: purchaserInfoData) as? [String: Any]
+            guard let purchaserInfoDict = maybeInfoDict, let info = PurchaserInfo(data: purchaserInfoDict) else {
+                return nil
+            }
+
+            if let schema = info.schemaVersion, schema == PurchaserInfo.currentSchemaVersion {
+                return info
+            }
+        } catch {
+            Logger.warn("Unable to unmarshall PurchaserInfo from cache:\n \(error.localizedDescription)")
+        }
+
+        return nil
+    }
+
+    @objc(cachePurchaserInfo:forAppUserID:)
+    public func cache(purchaserInfo: PurchaserInfo, appUserID: String) {
+        do {
+            let jsonData = try JSONSerialization.data(withJSONObject: purchaserInfo.jsonObject())
+            deviceCache.cache(purchaserInfo: jsonData, appUserID: appUserID)
+            sendUpdateIfChanged(purchaserInfo: purchaserInfo)
+        } catch {
+            Logger.warn("Invalid JSON returned from purchaserInfo.jsonObject\n\(error.localizedDescription)")
+        }
+    }
+
+    @objc public func clearPurchaserInfoCache(forAppUserID appUserID: String) {
+        purchaserInfoCacheLock.lock()
+        deviceCache.clearPurchaserInfoCache(appUserID: appUserID)
+        lastSentPurchaserInfo = nil
+        purchaserInfoCacheLock.unlock()
+    }
+
+    private func sendUpdateIfChanged(purchaserInfo: PurchaserInfo) {
+        guard let delegate = self.delegate else {
+            return
+        }
+
+        purchaserInfoCacheLock.lock()
+        guard lastSentPurchaserInfo != purchaserInfo else {
+            purchaserInfoCacheLock.unlock()
+            return
+        }
+
+        if lastSentPurchaserInfo != nil {
+            Logger.debug(Strings.purchaserInfo.sending_updated_purchaserinfo_to_delegate)
+        } else {
+            Logger.debug(Strings.purchaserInfo.sending_latest_purchaserinfo_to_delegate)
+        }
+
+        self.lastSentPurchaserInfo = purchaserInfo
+        operationDispatcher.dispatchOnMainThread {
+            self.purchaserInfoCacheLock.unlock()
```

Comment:
do we need to wait until we're on the main thread to release the lock? 
I'm somewhat concerned about switching threads with the lock

Reply:
The original implementation locks until the end of the method, but to be more technically correct, I think releasing the lock once the block is executed by the main thread gives use the extra reassurance that the call to `purchaserInfoManagerDidReceiveUpdated` happens in the same order as the change.